File: Solutions/Cpp/contains_duplicate_ii.cpp
/*
    [EASY]
    219. Contains Duplicate II

    Concepts:
    - array
    - hash table

    Stats:
        Runtime | 51 ms     [Beats 82.19%]
        Memory  | 81.31 MB  [Beats 27.01%]
*/

class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        std::unordered_map<int, int> idx = {};

        for (int i = 0; i < nums.size(); i++) {
            if (idx.contains(nums[i]) && std::abs(i - idx[nums[i]]) <= k)
                return true;
            idx[nums[i]] = i;
        }

        return false;
    }
};

File: Solutions/Rust/contains_duplicate_ii.rs
/*
    [EASY]
    219. Contains Duplicate II

    Concepts:
    - array
    - hash table

    Stats:
        Runtime | 17 ms     [Beats 27.40%]
        Memory  | 4.80 MB   [Beats 47.26%]
*/
use std::collections::HashMap;

impl Solution {
    pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
        let mut idx: HashMap<i32, i32> = HashMap::new();

        for i in 0..nums.len() {
            match idx.get(&nums[i]) {
                Some(&v) => {
                    if (i as i32 - v).abs() <= k {
                        return true;
                    }
                },
                None => {}
            }
            idx.insert(nums[i], i as i32);
        }

        return false;
    }
}